cdc/generation_id.hh

```diff
@@ -28,10 +28,17 @@
 
 namespace cdc {
 
-struct generation_id {
+struct generation_id_v1 {
     db_clock::time_point ts;
 };
 
+struct generation_id_v2 {
+    db_clock::time_point ts;
+    utils::UUID id;
+};
+
+using generation_id = std::variant<generation_id_v1, generation_id_v2>;
+
```

Comment:
Can we extend rather than add a variant? I think it will simplify many paths.

If we see a v1 generation id, append to it a fixed uuid (perhaps a null uuid, perhaps just some uuid we will generate once and hardcode). This will unify many code paths.

Reply:
For example, operator== for generation ids can be auto-generated with `operator<=>(const generation_id&) = default`.

Reply:
Having two separate types for v1 and v2 generation IDs protects us (in compile time) from introducing v2 generation IDs to the cluster by accident - for example, it would be very undesirable to introduce a v2 ID to a cluster in which some nodes do not yet understand the new format.

I want these different codepaths to be different, and I want the type system to force them to be different.

Reply:
How does it protect? We can still send a v1 where a v2 is expected.

Reply:
Functions which expect a v1 cannot get a v2 accidentally (`rewrite_streams_descriptions`, `cdc_set_rewritten`) because the types are different (and neither is a subtype of the other).

I also know where a v2 is first introduced: `make_new_cdc_generation` is the only function which returns a v2 and I can protect all places where this function is used with a feature check.

Which code paths do you think would be unified?

Reply:
Ok, I see. Hypothetically, if you introduced a new gossip application state for v2 you wouldn't need the variant? That is, the variant multiplexes the two types for transport on gossip and then de-multiplexes them on arrival?

Reply:
> Hypothetically, if you introduced a new gossip application state for v2 you wouldn't need the variant?

I don't strictly _need_ the variant. I could make v1 a special case of v2, but then I would have to be more careful not to accidentally send a v1 as a v2, for example. I just think that variant is an elegant method of protecting us from making a mistake.

Introducing a new gossip application state wouldn't change much in that regard. It could possibly allow us to perform upgrade without a cluster feature (because the existing gossip app state would always have v1 IDs and the new gossip app state would always have v2 IDs; non upgraded nodes would keep using existing gossip app state without risk of trying to read a v2 ID).

> That is, the variant multiplexes the two types for transport on gossip and then de-multiplexes them on arrival?

Yes. I could do it without a variant, for example:
```
struct generation_id {
    db_clock::time_point ts;
    std::optional<utils::UUID> uuid;
}
```
then when I insert the ID into gossip, I serialize it as v2 if and only if `uuid` is not nullopt (and otherwise I serialize it as v1).
Could also do without `optional` by using a special hardcoded `utils::UUID` but then I would have to distinguish v1 from v2 by comparing `uuid` field with this hardcoded value - if it's equal, serialize as v1, otherwise as v2.

Reply:
Even if I introduced a new gossip application state, I would still have to be careful not to send a v2 through the old appstate or a v1 through the new. The variant would give the same protection here as it does when we use a single appstate.